edgehog-device-runtime-containers/src/docker/image.rs

```diff
@@ -0,0 +1,680 @@
+
+        // Check if the image was not found
+        let inspect = match res {
+            Ok(inspect) => inspect,
+            Err(BollardError::DockerResponseServerError {
+                status_code: 404,
+                message,
+            }) => {
+                debug!("image not found: {message}");
+
+                return Ok(None);
+            }
+            Err(err) => return Err(ImageError::Inspect(err)),
+        };
+
+        if let Some(id) = &inspect.id {
+            self.id = Some(id.clone());
+        }
+
+        trace!("inspected image: {inspect:?}");
+
+        Ok(Some(inspect))
+    }
+
+    /// Remove an image and it's parents.
+    ///
+    /// See the [Docker API reference](https://docs.docker.com/engine/api/v1.43/#tag/Image/operation/ImageDelete)
+    #[instrument(skip_all, fields(image = self.id_or_ref()))]
+    pub async fn remove(
+        &self,
+        client: &Client,
+    ) -> Result<Option<Vec<ImageDeleteResponseItem>>, ImageError>
+    where
+        S: AsRef<str> + Debug + Display,
+    {
+        let id = self.id_or_ref();
+
+        debug!("removing {self}");
+
+        let res = client
+            .remove_image(id, None, self.docker_credentials()?)
+            .await;
+
+        match res {
+            Ok(delete_res) => {
+                info!("removed {self}");
+
+                Ok(Some(delete_res))
+            }
+            Err(BollardError::DockerResponseServerError {
+                status_code: 404,
+                message,
+            }) => {
+                warn!("image not found: {message}");
+
+                Ok(None)
+            }
+            Err(BollardError::DockerResponseServerError {
+                status_code: 409,
+                message,
+            }) => {
+                error!("cannot remove image in use: {}", &message);
+
+                // We need to unpack and repack the struct to access the fields
+                Err(ImageError::ImageInUse(
+                    BollardError::DockerResponseServerError {
+                        status_code: 409,
+                        message,
+                    },
+                ))
+            }
+            Err(err) => Err(ImageError::Remove(err)),
+        }
+    }
+}
+
+impl<S: Display> Display for Image<S> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Image")?;
+
+        if let Some(id) = &self.id {
+            write!(f, " ({id})")?;
+        }
+
+        write!(f, " {}", self.reference)
+    }
+}
+
+impl<S1, S2> PartialEq<Image<S2>> for Image<S1>
+where
+    S1: PartialEq<S2>,
+{
+    fn eq(
+        &self,
+        Image {
+            id,
+            reference,
+            registry_auth,
+        }: &Image<S2>,
+    ) -> bool {
+        let eq_registry_auth = match (&self.registry_auth, registry_auth) {
+            (Some(s1), Some(s2)) => *s1 == *s2,
+            (Some(_), None) | (None, Some(_)) => false,
+            (None, None) => true,
+        };
```

Comment:
```suggestion
        let eq_registry_auth = match (&self.registry_auth, registry_auth) {
            (Some(s1), Some(s2)) => *s1 == *s2,
            (None, None) => true,
            _ => false,
        };
```

Reply:
I'm wondering if even a simple `&self.registry_auth == registry_auth` would work

Reply:
I would prefer being explisit here in the match.

The PartialEq trait is implemented as `PartialEq<Option<T>> for Option<T>` while here we want to `PartialEq<Option<U>> for Option<T>`. So the `==` wont work.

Here the issue https://github.com/rust-lang/rust/issues/20063